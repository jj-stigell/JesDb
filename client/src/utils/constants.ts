import { Note, Profile } from 'src/types'

export const repoLink = 'https://github.com/jj-stigell/JesDb'

export const feedbackLink = 'https://forms.gle/SVNcSrujPHoy6UuG9'

export const ankiConnectDefaultAddress = 'http://127.0.0.1:8765'

// This one has highlighted_language set to 'ja' therefore it will have sentence_furigana_highlight
export const exampleNoteJa: Note = {
  id: 1863,
  sentence: '昔から 自動車の会社をやっていて',
  len: 16,
  sentence_furigana: '昔[むかし]から 自動車[じどうしゃ]の 会社[かいしゃ]をやっていて',
  translation: 'long time a go I had a car company',
  audio: `https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/gBumdOWWMhY/gBumdOWWMhY_s102_e104.mp3`,
  image: `https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/gBumdOWWMhY/gBumdOWWMhY_s102_e104.jpg`,
  start_time: 102,
  source_id: 'gBumdOWWMhY',
  highlighted_sentence: '昔から <span class="keyword">自動車</span>の会社をやっていて',
  highlighted_language: 'ja',
  sentence_furigana_highlight: '昔[むかし]から <span class="keyword">自動車</span>[じどうしゃ]の 会社[かいしゃ]をやっていて'
}

// This one has highlighted_language set to 'en' therefore sentence_furigana_highlight is empty
export const exampleNoteEn: Note = {
  id: 1135,
  sentence: 'それが怖くなったのでございます',
  len: 15,
  sentence_furigana: 'それが 怖[こわ]くなったのでございます',
  translation: 'I was scared of what was waiting for me.',
  audio: `https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/ylEimFz8KOo/ylEimFz8KOo_s159_e162.mp3`,
  image: `https://jynhfjtxpodsmjlnexrx.supabase.co/storage/v1/
  object/public/jesdb-media/ylEimFz8KOo/ylEimFz8KOo_s159_e162.jpg`,
  start_time: 159,
  source_id: 'ylEimFz8KOo',
  highlighted_sentence: 'I was <span class="keyword">scared</span> of what was waiting for me.',
  highlighted_language: 'en',
  sentence_furigana_highlight: ''
}

export const exampleProfile: Profile = {
  id: 1,
  name: 'Test Profile',
  deck_name: 'Test Deck',
  note_type: 'Basic',
  action_type: 'addNote',
  expression_field: 'expression',
  reading_field: 'reading',
  translation_field: 'translation',
  audio_field: 'audio',
  image_field: 'image',
  created_at: '2021-09-01T00:00:00.000Z',
  user_id: '1af8625c-0771-4c53-9851-42274e8e2385'
}
