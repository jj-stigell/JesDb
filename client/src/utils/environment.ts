/**
 * Current Node environment.
 * This variable indicates the environment in which the Node application is running.
 * Possible values: 'development', 'production', and 'test'.
 * Defaults to 'development' if the NODE_ENV environment variable is not set.
 */
export const NODE_ENV: string = process.env.REACT_APP_NODE_ENV ?? 'development'

/**
 * Supabase URL. This is the URL to your Supabase project.
 * You can find this key in your Supabase project settings.
 */
export const SUPABASE_URL: string = process.env.REACT_APP_SUPABASE_URL ?? ''

/**
 * Supabase API key. Only use public key here. Do not expose your private key.
 * This key is used to authenticate with the Supabase API.
 * You can find this key in your Supabase project settings.
 */
export const SUPABASE_KEY: string = process.env.REACT_APP_SUPABASE_KEY ?? ''

/**
 * Auth callback URL. This is the URL that the user will be redirected to after
 * they authenticate with Supabase.
 * This URL must be added to the list of "Redirect URIs" in your Supabase project settings.
 */
export const AUTH_CALLBACK_URL: string =
  process.env.REACT_APP_AUTH_CALLBACK_URL ?? 'http://localhost:3000/auth/callback'
