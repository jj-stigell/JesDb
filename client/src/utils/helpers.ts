import { NoteParams, Note, Profile } from 'src/types'

export function getFileName (url: string): string {
  const regex = /[^/]+\.(mp3|jpg|png)$/
  const match = url.match(regex)

  if (match !== null) {
    return match[0]
  } else {
    return `${(+new Date()).toString(36)}.${url.split('.').pop() ?? 'file'}`
  }
}

export function scrollToTop (): void {
  window.scrollTo({
    top: 0,
    behavior: 'smooth'
  })
}

export function constructParams (
  example: Note,
  profile: Profile
): NoteParams {
  const fields: any = {}
  const audio: any = {}
  const picture: any = {}

  if (profile.expression_field !== 'omit' && profile.expression_field !== null) {
    fields[profile.expression_field] = example.sentence
  }

  if (profile.reading_field !== 'omit' && profile.reading_field !== null) {
    fields[profile.reading_field] = example.sentence_furigana
  }

  if (profile.translation_field !== 'omit' && profile.translation_field !== null) {
    fields[profile.translation_field] = example.translation
  }

  if (profile.audio_field !== 'omit' && example.audio !== null && profile.audio_field !== null) {
    fields[profile.audio_field] = ''
    audio.url = example.audio
    audio.filename = getFileName(example.audio)
    audio.fields = [profile.audio_field]
  }

  if (profile.image_field !== 'omit' && example.image !== null && profile.image_field !== null) {
    fields[profile.image_field] = ''
    picture.url = example.image
    picture.filename = getFileName(example.image)
    picture.fields = [profile.image_field]
  }

  const params: any = {
    note: {
      fields,
      tags: ['JesDb', example.source_id],
      audio: [audio],
      picture: [picture]
    }
  }

  return params
}

export function parseRuby (input: string): string {
  const regex = /(^|\s|<[^>]+>)([^[<\s]+)\[([^\]]+)\]/g
  return input.replace(regex, (match, p1, kanji: string, furigana: string) => {
    const prefix: string = p1.trim().length > 0 ? p1 : ''
    return `${prefix}<ruby>${kanji}<rt>${furigana}</rt></ruby>`
  })
}
