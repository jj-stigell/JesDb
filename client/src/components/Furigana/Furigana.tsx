import React from 'react'
import { parseRuby } from '../../utils/helpers'

interface IFurigana {
  text: string
  withFurigana: string
  withHighlight: string
  highlightedLanguage: string
  withFuriganaAndHighlight: string
  showFurigana: boolean
}

export default function Furigana ({
  text,
  withFurigana,
  withHighlight,
  highlightedLanguage,
  withFuriganaAndHighlight,
  showFurigana
}: IFurigana): React.JSX.Element {
  const style = 'text-lg dark:text-white'

  if (!showFurigana) {
    if (highlightedLanguage === 'ja') {
      return (
        <div
          className={style}
          data-testid="furigana-component"
          lang="ja"
          aria-live="polite"
          dangerouslySetInnerHTML={{ __html: withHighlight }}
        />
      )
    } else {
      return (
        <div
          className={style}
          data-testid="furigana-component"
          aria-live="polite"
        >
          {text}
        </div>
      )
    }
  }

  const createMarkup = (): { __html: string } => {
    return {
      __html: parseRuby(withFurigana)
    }
  }

  return (
    <div
      className={style}
      data-testid="furigana-component"
      lang="ja"
      aria-live="polite"
      dangerouslySetInnerHTML={createMarkup()}
    />
  )
}
