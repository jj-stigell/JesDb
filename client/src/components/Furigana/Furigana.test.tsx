import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import Furigana from './Furigana'
import { exampleNoteJa } from '../../utils/constants'
import { parseRuby } from '../../utils/helpers'

const defaultProps = {
  text: exampleNoteJa.sentence,
  withFurigana: exampleNoteJa.sentence_furigana,
  withHighlight: exampleNoteJa.highlighted_sentence,
  highlightedLanguage: exampleNoteJa.highlighted_language,
  withFuriganaAndHighlight: exampleNoteJa.sentence_furigana_highlight,
  showFurigana: false
}

describe('Furigana Component', () => {
  it('renders highlighted text without furigana', () => {
    const { asFragment } = render(<Furigana {...defaultProps} />)

    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByTestId('furigana-component').innerHTML)
      .toContain(exampleNoteJa.highlighted_sentence)
  })

  it('renders non-highlighted text without furigana', () => {
    const { asFragment } = render(<Furigana {...defaultProps} highlightedLanguage='en' />)

    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByTestId('furigana-component').innerHTML)
      .toContain(exampleNoteJa.sentence)
  })

  it('does not render highlighted text with furigana even if showFurigana true', () => {
    const { asFragment } = render(<Furigana {...defaultProps} showFurigana={true} />)

    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByTestId('furigana-component').innerHTML)
      .toContain(parseRuby(exampleNoteJa.sentence_furigana))
  })

  it('renders non-highlighted text with furigana', () => {
    const { asFragment } =
      render(<Furigana {...defaultProps} showFurigana={true} highlightedLanguage='en' />)

    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByTestId('furigana-component').innerHTML)
      .toContain(parseRuby(exampleNoteJa.sentence_furigana))
  })
})
