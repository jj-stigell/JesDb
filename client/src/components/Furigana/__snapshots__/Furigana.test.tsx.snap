// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Furigana Component does not render highlighted text with furigana even if showFurigana true 1`] = `
<DocumentFragment>
  <div
    aria-live="polite"
    class="text-lg dark:text-white"
    data-testid="furigana-component"
    lang="ja"
  >
    <ruby>
      昔
      <rt>
        むかし
      </rt>
    </ruby>
    から
    <ruby>
      自動車
      <rt>
        じどうしゃ
      </rt>
    </ruby>
    の
    <ruby>
      会社
      <rt>
        かいしゃ
      </rt>
    </ruby>
    をやっていて
  </div>
</DocumentFragment>
`;

exports[`Furigana Component renders highlighted text without furigana 1`] = `
<DocumentFragment>
  <div
    aria-live="polite"
    class="text-lg dark:text-white"
    data-testid="furigana-component"
    lang="ja"
  >
    昔から 
    <span
      class="keyword"
    >
      自動車
    </span>
    の会社をやっていて
  </div>
</DocumentFragment>
`;

exports[`Furigana Component renders non-highlighted text with furigana 1`] = `
<DocumentFragment>
  <div
    aria-live="polite"
    class="text-lg dark:text-white"
    data-testid="furigana-component"
    lang="ja"
  >
    <ruby>
      昔
      <rt>
        むかし
      </rt>
    </ruby>
    から
    <ruby>
      自動車
      <rt>
        じどうしゃ
      </rt>
    </ruby>
    の
    <ruby>
      会社
      <rt>
        かいしゃ
      </rt>
    </ruby>
    をやっていて
  </div>
</DocumentFragment>
`;

exports[`Furigana Component renders non-highlighted text without furigana 1`] = `
<DocumentFragment>
  <div
    aria-live="polite"
    class="text-lg dark:text-white"
    data-testid="furigana-component"
  >
    昔から 自動車の会社をやっていて
  </div>
</DocumentFragment>
`;
